src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs

```diff
@@ -1429,5 +1429,135 @@ internal ImmutableArray<BoundExpression> MakeTempsForDiscardArguments(ImmutableA
+
+            TypeSymbol boolType = Compilation.GetSpecialType(CodeAnalysis.SpecialType.System_Boolean);
+
+            // Fold compile-time comparisons.
+            if (rewrittenExpr.ConstantValue != null)
+            {
+                switch (operatorKind)
+                {
+                    case BinaryOperatorKind.Equal:
+                        return Literal(ConstantValue.Create(rewrittenExpr.ConstantValue.IsNull, ConstantValueTypeDiscriminator.Boolean), boolType);
+                    case BinaryOperatorKind.NotEqual:
+                        return Literal(ConstantValue.Create(rewrittenExpr.ConstantValue.IsNull, ConstantValueTypeDiscriminator.Boolean), boolType);
+                }
+            }
+
+            TypeSymbol objectType = SpecialType(CodeAnalysis.SpecialType.System_Object);
+
+            if ((object)exprType != null)
+            {
+                if (exprType.Kind == SymbolKind.TypeParameter)
+                {
+                    // Box type parameters.
+                    rewrittenExpr = Convert(objectType, rewrittenExpr, Conversion.Boxing);
+
+                }
+                else if (exprType.IsNullableType())
+                {
+                    operatorKind |= BinaryOperatorKind.NullableNull;
+                }
+            }
+            if (operatorKind == BinaryOperatorKind.NullableNullEqual || operatorKind == BinaryOperatorKind.NullableNullNotEqual)
+            {
+                return RewriteNullableNullEquality(syntax, operatorKind, rewrittenExpr, Literal(ConstantValue.Null, objectType), boolType); 
+            }
+            else
+            {
+                return Binary(operatorKind, boolType, rewrittenExpr, Null(objectType));
+            }
+        }
+
+        internal BoundExpression MakeNullableHasValue(SyntaxNode syntax, BoundExpression expression)
+        {
+            return BoundCall.Synthesized(syntax, expression, LocalRewriter.UnsafeGetNullableMethod(syntax, expression.Type, CodeAnalysis.SpecialMember.System_Nullable_T_get_HasValue, Compilation, Diagnostics));
+        }
+
+        internal BoundExpression RewriteNullableNullEquality(
+            SyntaxNode syntax,
+            BinaryOperatorKind kind,
+            BoundExpression loweredLeft,
+            BoundExpression loweredRight,
+            TypeSymbol returnType)
+        {
+            // This handles the case where we have a nullable user-defined struct type compared against null, eg:
+            //
+            // struct S {} ... S? s = whatever; if (s != null)
+            //
+            // If S does not define an overloaded != operator then this is lowered to s.HasValue.
+            //
+            // If the type already has a user-defined or built-in operator then comparing to null is
+            // treated as a lifted equality operator.
+
+            Debug.Assert(loweredLeft != null);
+            Debug.Assert(loweredRight != null);
+            Debug.Assert((object)returnType != null);
```

Comment:
use the `(object)expr != null` pattern for the rest of these?

Reply:
Dismissed due to not-my-code